### ISSUE TITLE
panic: VersionUnavailable: Requested OpenGL version 4.3, got version 4.1

### ISSUE THREAD

**Issue by ghost** [MauriceGit/RealHeightmapTerrain#2]
./HeightmapTerrain 
panic: VersionUnavailable: Requested OpenGL version 4.3, got version 4.1

i am on a Mac. I did not know that the opengl version you compile against means you also need that version supported by you graphics cad. Is this true ?

**Comment by MauriceGit** [comment]
Yes, this is true. In this case, OpenGL 4.3 is required because of the use of compute shaders for calculating normals on the GPU directly.

If I remember correctly, I don't actually use the computed normals at the moment (8bit precision too low), so it should be relatively easy to remove them and lower the OpenGL core version to 4.1 or even down to 3.3 (possibly).

If you need assistance for that, I can point you to some parts of the code.

**Comment by ghost** [comment]
Hey..

ah thanks for explaining that. opengl is hard...
I tried changing the import paths to use 4.1 everywhere. It compiles, but then at runtime i get the same error. 
Mhhh.... i will try 3.3 now and see if i get the same error

---

this all leads to a YIKES realisation :). You have to compile many versions of your code against many versions of potential opengl drivers that users might have ???

**Comment by MauriceGit** [comment]
Hi,

just changing the import will not work as long as the code still uses features and functionality from OpenGL 4.3 or up.

Basically, the Go OpenGL library is just a foreign function interface to C, which calls the graphics drivers directly. 

Right now, functions like gl.DispatchCompute() and one/two more are in use, that are not available in OpenGL 4.1.

In production, you define the minimum required OpenGL version for your code. Then define lets say OpenGL 3.3 (select that in code. It's good practise to do that but you can also just skip it and let GLFW take the most current one).

From OpenGL 3.0 onwards, OpenGL is backward compatible. So function pointers stay the same. Code written against OpenGL 3.3 will work just fine, if your Graphics drivers support OpenGL 4.6. But not the other way around.

**Comment by ghost** [comment]
Thanks so much for explaining that. Its now makes sense.

Its a relief to hear that OpenGL 3.0 is backwards compatible. So even a Raspi Pi might work for example.

SO for things that require more advanced opengl ( for example functions like gl.DispatchCompute() ) would it make sense to use Build tags ?

**Comment by MauriceGit** [comment]
Yes, OpenGL works on the Pi. In this case it is just important not to use features from newer OpenGL versions.
In OpenGL versions < 3.0 we had the fixed-function pipeline, which is vastly outdated and rightfully deprecated. Most graphics drivers will still support that but are not guaranteed to keep supporting it. So it should be avoided at all costs.

Normally, the user is responsible for configuring the graphics drivers. This is not something you (as a developer shipping applications) can do (Other than pointing to the e.g. Nvidia homepage).

If you don't tell your application specifically, which OpenGL version to use, it will default to the highest version the corresponding system offers. 
So no build tags should be required normally :)

Edit: 
To make it a little more complicated - There is OpenGL ES (embedded system). It's version numbers differ. OpenGL ES 3.1 roughly corresponds to OpenGL 4.3 for desktops.
OpenGL ES is basically a subset of normal desktop OpenGL. Functions and general functionality are the same. But some things are missing. So some extra care is required when porting/writing OpenGL ES.
This is done, to keep OpenGL ES a little more slim and a enforce some coding style that is more efficient.
This also means that OpenGL ES will run on Desktop hardware but not necessarily the other way around.

The Raspberry Pi 3 should be able to support OpenGL ES 2.0, I think.

**Comment by ghost** [comment]
thanks for the explanation.
My MaxOS did not revert to the highest version that it supported.
It just refused to run. That was the original reason for raising this issue.
SO not sure whats wrong here.... what do you think..

**Comment by MauriceGit** [comment]
Probably the use of functionality >= OpenGL 4.1. Meaning most likely the call to 
gl.DispatchCompute(wgSizeX, wgSizeY, 1);
gl.MemoryBarrier(gl.SHADER_IMAGE_ACCESS_BARRIER_BIT);
Try commenting both lines and see if it works.
